pennylane_sf/gbs.py

```diff
@@ -0,0 +1,258 @@
+        "ParamGraphEmbed": GraphEmbed,
+    }
+
+    _observable_map = {
+        "Identity": identity,
+    }
+
+    _circuits = {}
+
+    _capabilities = {"model": "cv", "provides_jacobian": True}
+
+    def __init__(self, wires, *, analytic=True, cutoff_dim, shots=1000):
+        super().__init__(wires, analytic=analytic, shots=shots)
+        self.cutoff = cutoff_dim
+        self._params = None
+        self._WAW = None
+
+    @staticmethod
+    def calculate_WAW(params, A, n_mean):
+        """Calculates the :math:`WAW` matrix.
+
+        Rescales :math:`A` so that when encoded in GBS the mean photon number is equal to
+        ``n_mean``.
+
+        Args:
+            params (array[float]): variable parameters
+            A (array[float]): adjacency matrix
+            n_mean (float): mean number of photons
+
+        Returns:
+            array[float]: the :math:`WAW` matrix
+        """
+        A *= rescale(A, n_mean)
+        W = np.diag(np.sqrt(params))
+        return W @ A @ W
+
+    @staticmethod
+    def calculate_n_mean(A):
+        """Calculates the mean number of photons for an adjacency matrix encoded into GBS.
+
+        Note that for ``A`` to be directly encoded into GBS, its singular values must not exceed
+        one.
+
+        Args:
+            A (array[float]): adjacency matrix
+
+        Returns:
+            float: mean number of photons
+        """
+        singular_values = np.linalg.svd(A, compute_uv=False)
+        return np.sum(singular_values ** 2 / (1 - singular_values ** 2))
+
+    # pylint: disable=missing-function-docstring
+    def execute(self, queue, observables, parameters=None, **kwargs):
+        parameters = parameters or {}
+        if len(queue) > 1:
+            raise ValueError(
+                "The StrawberryFieldsGBS device accepts a single application of ParamGraphEmbed"
+            )
+        return super().execute(queue, observables, parameters, **kwargs)
+
+    # pylint: disable=pointless-statement,expression-not-assigned
+    def apply(self, operation, wires, par):
+        self._params, A, _ = par
+
+        if len(self._params) != self.num_wires:
+            raise ValueError(
+                "The number of variable parameters must be equal to the total number of wires."
+            )
+
+        self._WAW = self.calculate_WAW(*par)
+        n_mean_WAW = self.calculate_n_mean(self._WAW)
+
+        op = GraphEmbed(self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
+        op | [self.q[wires.index(i)] for i in wires]
+
+        if not self.analytic:
+            MeasureFock() | [self.q[wires.index(i)] for i in wires]
+
+    def pre_measure(self):
+        self.eng = sf.Engine("gaussian", backend_options={"cutoff_dim": self.cutoff})
+
+        if self.analytic:
+            results = self.eng.run(self.prog)
+        else:
+            results = self.eng.run(self.prog, shots=self.shots)
+
+        self.state = results.state
+        self.samples = results.samples
```

Comment:
Question: now that we are locked into _only_ using the Gaussian backend, why not go straight to The Walrus and simply use either `thewalrus.samples.hafnian_sample_graph` (non-analytic mode) or `thewalrus.quantum.probabilities` (analytic mode)? This should remove significant overhead.

The only advantage to going through the SF frontend that I can see is:

* The ability to hot-swap backends
* If you want to take advantage of compilation, e.g., if you are running on a hardware device (not yet supported).